feature-libs/product-configurator/rulebased/root/interceptor/cpq-access-storage.service.ts

```diff
@@ -0,0 +1,132 @@
+  protected readonly EXPIRED_TOKEN = {
+    accessToken: 'INVALID DUMMY',
+    accessTokenExpirationTime: 0,
+  };
+
+  constructor(
+    protected cpqAccessLoaderService: CpqAccessLoaderService,
+    protected authService: AuthService,
+    protected config: CpqConfiguratorAuthConfig
+  ) {}
+
+  protected cpqAccessObservable: Observable<CpqAccessData>;
+  protected currentCpqAccessSubscription: Subscription;
+  protected cpqAccessDataCache: BehaviorSubject<CpqAccessData>;
+
+  protected currentCpqSessionId: string | null = null;
+
+  getCachedCpqAccessData(): Observable<CpqAccessData> {
+    if (!this.cpqAccessObservable) {
+      this.initCpqAccessObservable();
+    }
+    return this.cpqAccessObservable;
+  }
+
+  /**
+   * Renews the current access data. All subscribers of getCachedCpqAccessData()
+   * will receive the new data. Will only have an effect, if there are any subscribers
+   * and the user is logged in.
+   */
+  renewCachedCpqAccessData() {
+    // only force token refresh if initialized.
+    if (this.cpqAccessObservable) {
+      this.stopAutoFetchingCpqAccessData();
+      this.cpqAccessDataCache.next(this.EXPIRED_TOKEN); // invalidate cache
+      this.authService
+        .isUserLoggedIn()
+        .pipe(take(1)) // get current login state
+        .subscribe((loggedIn) => {
+          // only fetch new token if user is logged in.
+          if (loggedIn) {
+            this.startAutoFetchingCpqAccessData();
+          }
+        });
+    }
+  }
+
+  protected initCpqAccessObservable() {
+    this.cpqAccessDataCache = new BehaviorSubject(this.EXPIRED_TOKEN);
+    this.cpqAccessObservable = this.cpqAccessDataCache.pipe(
+      // Never expose expired tokens - either cache was invalidated with expired token,
+      // or the cached one expired before a new one was fetched.
+      filter((data) => !this.isTokenExpired(data))
+    );
+    this.authService
+      .isUserLoggedIn()
+      .pipe(distinctUntilChanged()) // only react if user login status changes
+      .subscribe((loggedIn) => {
+        if (loggedIn) {
+          // user logged in - start/stop to ensure token is refreshed
+          this.stopAutoFetchingCpqAccessData();
+          this.startAutoFetchingCpqAccessData();
+        } else {
+          // user logged of - cancel token fetching
+          this.stopAutoFetchingCpqAccessData();
+          this.cpqAccessDataCache.next(this.EXPIRED_TOKEN); // invalidate cache
+        }
+      });
+  }
+
+  protected stopAutoFetchingCpqAccessData() {
+    this.currentCpqAccessSubscription?.unsubscribe();
+  }
+
+  protected startAutoFetchingCpqAccessData() {
+    this.currentCpqAccessSubscription = this.cpqAccessLoaderService
+      .getCpqAccessData()
+      .pipe(
+        expand((data) =>
+          timer(this.fetchNextTokenIn(data)).pipe(
+            switchMap(() => this.cpqAccessLoaderService.getCpqAccessData())
+          )
+        )
+      )
+      .subscribe(this.cpqAccessDataCache); // also propagate errors
```

Comment:
Let's say that OCC call to `getCpqAccessData` returns an error (connection issue, whatever...). If I understand correctly above code, then:
1. `this.currentCpqAccessSubscription` will be stopped 
2.  `this.cpqAccessDataCache` behavior subject will get the propagated error

Regarding 2. Please note that BehaviorSubject propagated with an error is disabled and cannot emit any new normal values. I guess you'd like to `catchError()` inside switchMap and handle it somehow. Please note in line 92 errors are also not catched.

Regarding 1. I have a gut feeling that dying the subscription `currentCpqAccessSubscription` on the errored http call might break the user experience (a page refresh is needed). Please discuss in your team how you'd like to handle such a case? Show an error message? Retry one or 2 times, and only then show an error message?

You might want to simulate the error by providing a custom http interceptor that will catch calls to OCC endpoint `getCpqAccessData` and return error on some condition.

I did only mimic it the above flow in a simple stackblitz https://stackblitz.com/edit/error-behavior-subject?file=index.ts Please test it in your app

Reply:
created https://github.com/SAP/spartacus/issues/12036 for it

Reply:
@Platonn Not sure if we can provide any meaningful handling here. If we can't obtain a token, we can't communicate with CPQ and configuration is not possible.
In theory you could have something like an offline mode were you can continue to click around and then do a bunch of requests as soon as you have connectivity again. Could be useful for mobile sales persons, which do not have always stable connection.
BUT - This is a definitely an own Backlog item / user story and needs to be designed and crafted carefully. There a lot of pitfalls with that approach.

Reply:
So - I just created  minimal solution for now, that ensures that a new token can be fetched after an error happened in case you re-enter the configuration screen. implemented by extending the check when to init the token data
```
if (!this.cpqAccessData$ || this._cpqAccessData$.hasError) {
      this.initCpqAccessData();
```